packages/graph/hash_graph/lib/graph/src/api/rest/utoipa_typedef/subgraph/edges.rs

```diff
@@ -56,13 +81,68 @@ impl From<crate::subgraph::edges::Edges> for Edges {
             knowledge_graph: KnowledgeGraphRootedEdges(edges.knowledge_graph.into_iter().fold(
                 HashMap::new(),
                 |mut map, (id, edges)| {
-                    let edges = edges.into_iter().collect();
+                    let edges = edges.into_iter().map(|edge| {
+                        match edge {
+                            crate::subgraph::edges::KnowledgeGraphOutwardEdges::ToOntology(edge) => {
+                                KnowledgeGraphOutwardEdges::ToOntology(edge)
+                            }
+                            crate::subgraph::edges::KnowledgeGraphOutwardEdges::ToKnowledgeGraph(
+                                edge,
+                            ) => {
+                                // We avoid storing redundant information when multiple editions of
+                                // the endpoints or links are present and in order to easily look
+                                // the corresponding link up in the vertices we store the earliest
+                                // timestamp when a link was added to the entity.
+                                //
+                                // As the vertices are sorted by timestamp, it's possible to get all
+                                // vertices starting with the provided earliest timestamp without
+                                // further filtering.
+                                //
+                                // We have four different permutations of a knowledge-knowledge
+                                // edge:
+                                //   1. `HAS_LEFT_ENTITY`, `reversed = false`
+                                //   2. `HAS_RIGHT_ENTITY`, `reversed = false`
+                                //   3. `HAS_LEFT_ENTITY`, `reversed = true`
+                                //   4. `HAS_RIGHT_ENTITY`, `reversed = true`
+                                //
+                                // For 1. and 2., the entity is a link and we want to store the
+                                // earliest version of this link as the earliest timestamp.
+                                // For 3. and 4., the endpoint of the edge is a link and we want to
+                                // store the earliest of that endpoint as the earliest timestamp.
```

Comment:
Is it correct that we are using the `right_endpoint` for both the `LEFT` and `RIGHT` permutation of `edge.reversed = true`? Or should we use a different endpoint for them?

Reply:
Yes, that is correct. The "right_endpoint" refers to the edge itself, not the edge type.